[Console] Bind input before executing the COMMAND event

wouterj opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #10695 (problem 1) |
| License | MIT |
| Doc PR | - |

Previously, `$input` wasn't very usefull in the `console.command` event, as the input was not yet bound to the command definition.

With this PR, the input is now bound twice: Once before the event is dispatched (to make it usefull in the listeners) and once at the original location in `Command#run()` (to allow changing the input definition in an event listener).

wouterj committed:
Bind input before executing the COMMAND event

stof commented:
wouldn't this break cases where the command is allowing to get arguments interactively in ``interact()`` ?

WouterJ commented:
@stof afaics, this only binds, the validation is still done after interact. So I don't think it should cause trouble.

It also means listeners can add options/arguments without getting validation errors.

WouterJ commented:
Failures related to https://github.com/symfony/symfony/pull/15940 and not this PR

xabbuh commented:
@WouterJ You could rebase as #15940 was merged (though we will still see some failures due to the security tests I think).

Tobion commented:
Please also add a test where the listener adds an option. And check if it can be used.

Tobion commented:
I was looking when validation actually happens as I saw symfony/console#15.
It seems validation can happen both 
- in `$input->bind` (when parsing), see https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Console/Command/Command.php#L234
- in `$input->validate();` which is called after `interact`, see https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Console/Command/Command.php#L256

Tobion commented:
> wouldn't this break cases where the command is allowing to get arguments interactively in interact() ?

So this case from @stof doesn't break because
- wrong arguments/options, e.g. too many would already error out when binding anyway
- missing arguments can still be added in interact as the number of arguments is only validated later

Tobion commented:
> Please also add a test where the listener adds an option. And check if it can be used.

I think exactly that is still no possible with this solution. If you call a command with an option/argument that is only about to be added by a listener, it will throw an exception already when trying to parse the input via `$input->bind()`. So the listener won't have a chance to add change the definition.

Tobion commented:
I think to solve the problem, we really need to refactor the validation. `bind()` should not validate but should parse everything indepently. And only `validate` checks if there are too many/missing/wrong arguments or options.

WouterJ commented:
@Tobion yeah, or wrap things inside a try..catch loop as is done in other bind cases.

WouterJ commented:
So `Input#validate()` only validates if required arguments are passed. `Input#bind()` validates if passed options/arguments are valid. Added a try..catch loop around the binding before the event listener and added a test.

Tobion commented on src/Symfony/Component/Console/Application.php:
```diff
@@ -880,6 +882,16 @@ protected function doRunCommand(Command $command, InputInterface $input, OutputI
+        try {
+            $command->mergeApplicationDefinition();
+            $input->bind($command->getDefinition());
+        } catch (InvalidArgumentException $e) {
```
I think it would be easiest to just catch all `Symfony\Component\Console\Exception\ExceptionInterface`

Tobion replied:
similarly https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/Console/Command/Command.php#L234 should be updated

WouterJ commented:
ping @symfony/deciders PR is now updated and ready to be merged imo

Tobion commented:
:+1: 

Status: Reviewed

xabbuh commented:
:+1:

Should we add a chanelog entry so that people interested in changing the definition notice this is possible now?

fabpot commented:
Thank you @WouterJ.
